src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -441,11 +461,131 @@ internal IEnumerable<IEdmNavigationProperty> GetSelectedNavigationProperties(IEd
             return selectedNavigationProperties.Distinct();
         }
 
+        /// <summary>
+        /// Gets all selected structured properties at the current node.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The selected structured properties at this node level.</returns>
+        /// <remarks>Result does not include the dynamic properties, which can be retrieved by <code>GetSelectedDynamicProperties</code></remarks>
+        internal IEnumerable<IEdmProperty> GetSelectedProperties(IEdmStructuredType structuredType)
+        {
+            if (this.selectionType == SelectionType.Empty)
+            {
+                return EmptyEdmProperties;
+            }
+
+            // We cannot determine the selected properties without the user model. This means we won't be computing the missing properties.
+            if (structuredType == null)
+            {
+                return EmptyEdmProperties;
+            }
+
+            if (this.selectionType == SelectionType.EntireSubtree || this.hasWildcard)
+            {
+                return structuredType.Properties();
+            }
+
+            Debug.Assert(this.selectedProperties != null, "selectedProperties != null");
+
+            // Get properties selected, and filter out unrecognized properties.
+            return this.selectedProperties.Select(structuredType.FindProperty).OfType<IEdmProperty>();
+        }
+
+        /// <summary>
+        /// Return names of dynamic properties.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The names of dynamic properties.</returns>
+        internal IEnumerable<string> GetSelectedDynamicProperties(IEdmStructuredType structuredType)
+        {
+            IEnumerable<string> dynamicProperties = Enumerable.Empty<string>();
+
+            if (this.selectedProperties != null && this.selectedProperties.Count > 0)
+            {
+                dynamicProperties = this.selectedProperties.Select(
+                    prop => structuredType.FindProperty(prop) == null
+                    ? prop
+                    : null)
+                    .OfType<string>();
+            }
+
+            return dynamicProperties;
+        }
+
+        /// <summary>
+        /// Gets the list of names for selected declared or dynamic properties having values of null.

```

Comment:
> Gets the list of names for selected declared or dynamic properties having values of null. [](start = 11, length = 90)

Does this return the names of the properties having null values, or the names of (nullable) properties missing from the payload?

Reply:
Updated. Should be the names of properties missing from payload, which will be restored to null. 

---
In reply to: [233188777](https://github.com/OData/odata.net/pull/1106#discussion_r233188777) [](ancestors = 233188777)